src/lib/OS/Process.cc

```diff
@@ -98,52 +140,69 @@ Process::Process(const std::vector<std::string>& commandLine,
-  uint64_t heapStart = alignToBoundary(instructions.size(), 32);
-
-  // Set mmap region start to be an equal distance from the stack and heap
-  // starts. Additionally, align to the page size (4kb)
-  uint64_t mmapStart =
-      alignToBoundary(heapStart + (heapSize + stackSize) / 2, pageSize_);
-
-  // Calculate process image size, including heap + stack
-  uint64_t size = heapStart + heapSize + stackSize;
-  // Check if global memory size is greater than process image size.
-  if (memory->getMemorySize() < size) {
-    std::cerr << "[SimEng:Process] Memory size is less than size of the "
-                 "process image. Please "
-                 "increase memory size"
-              << std::endl;
-    std::exit(1);
-  }
-
-  char* unwrappedProcImgPtr = (char*)malloc(size * sizeof(char));
-  std::copy(instructions.begin(), instructions.end(), unwrappedProcImgPtr);
-
-  uint64_t stackPtr = createStack(&unwrappedProcImgPtr, size);
-  memRegion_ = MemRegion(stackSize, heapSize, size, stackPtr, heapStart,
-                         pageSize_, mmapStart);
-  // copy process image to global memory.
-  memory->sendUntimedData(unwrappedProcImgPtr, 0, size);
-  fileDescriptorTable_.emplace_back(STDIN_FILENO);
-  fileDescriptorTable_.emplace_back(STDOUT_FILENO);
-  fileDescriptorTable_.emplace_back(STDERR_FILENO);
-  free(unwrappedProcImgPtr);
+  uint64_t mmapSize = config["Process-Image"]["Mmap-Size"].as<uint64_t>();
+
+  uint64_t instrSize = upAlign(instructions.size(), PAGE_SIZE);
+  uint64_t instrEnd = instrSize;
+
+  // Heap grows upwards towards higher addresses.
+  heapSize = upAlign(heapSize, PAGE_SIZE);
+  uint64_t heapStart = instrEnd + PAGE_SIZE;
+  uint64_t heapEnd = heapStart + heapSize;
+
+  // Mmap grows upwards towards higher addresses.
+  mmapSize = upAlign(mmapSize, PAGE_SIZE);
+  uint64_t mmapStart = heapEnd + PAGE_SIZE;
+  uint64_t mmapEnd = mmapStart + mmapSize;
+
+  // Stack grows downwards towards lower addresses.
+  stackSize = upAlign(stackSize, PAGE_SIZE);
+  uint64_t stackEnd = mmapEnd + PAGE_SIZE;
+  uint64_t stackStart = stackEnd + stackSize;
+  uint64_t size = stackStart;
+
+  // Request Page frames for heap and stack memory.
+  uint64_t instrPhyAddr = OS_->requestPageFrames(instrSize);
+  uint64_t heapPhyAddr = OS_->requestPageFrames(heapSize);
+  uint64_t stackPhyAddr = OS_->requestPageFrames(stackSize);
+
+  // Create page table mappings for stack and heap virtual address ranges.
+  pageTable_->createMapping(0, instrPhyAddr, instrSize);
+  pageTable_->createMapping(heapStart, heapPhyAddr, heapSize);
+  pageTable_->createMapping(stackEnd, stackPhyAddr, stackSize);
+  uint64_t stackPtr = createStack(stackStart);
+
+  // Create the callback function which will be used by MemRegion to unmap page
+  // table mappings upon VMA deletes.
+  std::function<uint64_t(uint64_t, size_t)> unmapFn =
+      [this](uint64_t vaddr, size_t size) -> uint64_t {
+    return pageTable_->deleteMapping(vaddr, size);
+  };
```

Comment:
Why does this function definition differ to the one in the other constructor? The other one does a check on the value returned from `pageTable_deleteMapping()` before it is returned

Reply:
done